File: PICHY/Utils/FiscalYear.swift
//
//  FiscalYear.swift
//  PICHY
//
//  Thai government fiscal year: 1 October – 30 September.
//  e.g. ปีงบประมาณ 2569 runs 1 Oct 2025 (พ.ศ. 2568) – 30 Sep 2026 (พ.ศ. 2569).
//

import Foundation

enum FiscalYear {
    private static let startMonth = 10  // October

    /// The half-open-ish [start, end] Gregorian range containing `date`.
    static func range(containing date: Date) -> (start: Date, end: Date) {
        let cal = Calendar.gregorian
        let comps = cal.dateComponents([.year, .month], from: date)
        let year = comps.year ?? 2025
        let month = comps.month ?? 1
        let startYear = month >= startMonth ? year : year - 1
        let start = cal.date(from: DateComponents(year: startYear, month: startMonth, day: 1))!
        let end = cal.date(from: DateComponents(year: startYear + 1, month: startMonth, day: 1))!
            .addingTimeInterval(-1)
        return (start, end)
    }

    /// Buddhist-era fiscal year number, e.g. 2569.
    static func buddhistYear(containing date: Date) -> Int {
        let cal = Calendar.gregorian
        let comps = cal.dateComponents([.year, .month], from: date)
        let year = comps.year ?? 2025
        let month = comps.month ?? 1
        let endGregYear = month >= startMonth ? year + 1 : year
        return endGregYear + 543
    }

    static func label(containing date: Date) -> String {
        "ปีงบประมาณ \(buddhistYear(containing: date))"
    }

    static func contains(_ date: Date, sameFiscalYearAs anchor: Date) -> Bool {
        let r = range(containing: anchor)
        return date >= r.start && date <= r.end
    }
}

File: web/src/utils/fiscalYear.ts
// Thai government fiscal year: 1 Oct – 30 Sep.
import { fromKey } from './thaiDate'

const START_MONTH = 9 // October (0-based)

export function fiscalRange(key: string): { start: string; end: string } {
  const d = fromKey(key)
  const year = d.getFullYear()
  const startYear = d.getMonth() >= START_MONTH ? year : year - 1
  const start = new Date(startYear, START_MONTH, 1)
  const end = new Date(startYear + 1, START_MONTH, 0) // 30 Sep next year
  return { start: keyOf(start), end: keyOf(end) }
}

export function sameFiscalYear(key: string, anchor: string): boolean {
  const r = fiscalRange(anchor)
  return key >= r.start && key <= r.end
}

export function fiscalBuddhistYear(key: string): number {
  const d = fromKey(key)
  const endGreg = d.getMonth() >= START_MONTH ? d.getFullYear() + 1 : d.getFullYear()
  return endGreg + 543
}

export function fiscalLabel(key: string): string {
  return `ปีงบประมาณ ${fiscalBuddhistYear(key)}`
}

function keyOf(date: Date): string {
  const y = date.getFullYear()
  const m = String(date.getMonth() + 1).padStart(2, '0')
  const d = String(date.getDate()).padStart(2, '0')
  return `${y}-${m}-${d}`
}
